C#: mass enable diff-informed data flow

d10c opened the pull request:
An auto-generated patch that enables diff-informed data flow in the obvious cases.

Builds on https://github.com/github/codeql/pull/18344 and https://github.com/github/codeql-patch/pull/88

d10c commented:
It turns out that some of the generated changes in the PRs were not correct, e.g. because they should have also generated a `getASelected{Source,Sink}Location()` override but didn't ([see Chuan-kai's comment here](https://github.com/github/codeql-team/issues/3926#issuecomment-2944599988)). So for now I'm putting them back in Draft until I make sure (via the patch script) that we are correctly handling all 3 [documented query patterns](https://github.com/github/semmle-code/blob/main/documentation/markdown-internal/incremental-codeql.md#simple-dataflow-query), starting with the simplest one (both source and sink are used as location sources). If you have already started reviewing the PRs, thank you (also for your patience) and stay tuned for an update as to what has changed in the meantime!

michaelnebel commented:
@d10c : Great!
Thank you for doing this. A couple of questions
- Do you expect there will be any configurations that will not be diff-informed?
- How does it impact tests (I remember hearing something along the lines of that tests already worked with "diff-informed", if the tests are `.qlref` files)?

d10c commented:
**Update: no changes since last time I opened the PR.** It turns out that it's sound (but not optimally performant) to leave `getASelected{Source,Sink}Location()` un-overridden, specifically in case of a select clause containing only one of source or sink but not both. The patch script currently does not differentiate between that case and the one in which both source and sink are present in the select clause. So I will re-open these PRs as they are, and generate an appropriate `getASelected{Source,Sink}Location()` override in a follow-up round of PRs.

d10c commented:
> * Do you expect there will be any configurations that will not be diff-informed?

In this initial PR, we are skipping over the cases where a node other than source or sink is used as a location source in the select clause. To enable these cases, a non-empty override of `getASelected{Source,Sink}Location()` will be needed. Also, queries with more than one dataflow config are probably another case to be considered separately.

> * How does it impact tests (I remember hearing something along the lines of that tests already worked with "diff-informed", if the tests are `.qlref` files)?

It's true that in order to enable diff-informed testing of queries, the tests have to be .qlref files. I'm planning on doing that kind of rewrite (similar to [this PR](https://github.com/github/codeql/pull/18848/files#top)) in a later phase.

michaelnebel commented:
> It's true that in order to enable diff-informed testing of queries, the tests have to be .qlref files. I'm planning on doing that kind of rewrite (similar to [this PR](https://github.com/github/codeql/pull/18848/files#top)) in a later phase.

Fantastic! We are also rewriting the test cases whenever we touch the queries (to use inline expectations) - so that would be great! 😄

d10c commented on csharp/ql/lib/change-notes/2025-06-04-diff-informed.md:
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* A number of built-in C# queries can now run in diff-informed mode.
```
No, it hasn't been made public. So it probably makes sense to not have a release note on any of these PRs.

d10c commented on csharp/ql/lib/semmle/code/csharp/security/dataflow/SqlInjectionQuery.qll:
```diff
@@ -43,6 +43,8 @@ module SqlInjectionConfig implements DataFlow::ConfigSig {
    */
   predicate isBarrier(DataFlow::Node node) { node instanceof Sanitizer }
+
+  predicate observeDiffInformedIncrementalMode() { any() }
```
Since (AFAIK) PathNode and Node have the same location, it shouldn't make a difference.

d10c commented:
As for DCA, it looks like we don't have sources set up with diffs, so we can't use DCA to test performance on diffs yet. A normal DCA run would then test before-and-after-this-PR without diff-informed mode, but those two runs should have the same behaviour. So I say we leave DCA testing out for now until I get the diff-informed sources in place.

d10c commented:
Note, according to the [follow-up PR](https://github.com/github/codeql/pull/19758), one of the queries ([csharp/ql/src/Security Features/CWE-327/DontInstallRootCert.ql](https://github.com/github/codeql/pull/19758/commits/9fde7ceff6a80287b7f7fa927a6780c21e5398dd)) has a missing source/sink location in its select clause; the other ones should have both.
